apps/class-solid/src/components/ajvForm.tsx

```diff
@@ -0,0 +1,43 @@
+import type {
+  FieldValues,
+  FormErrors,
+  PartialValues,
+  ValidateForm,
+} from "@modular-forms/solid";
+import type { ValidateFunction } from "ajv";
+
+/**
+ * Creates a modular form validator from an AJV validate function.
+ *
+ * @param validate - The AJV validate function. Return of `new Ajv().compile(schema)`.
+ * @returns A validation function.
+ */
+export function ajvForm<TFieldValues extends FieldValues>(
+  validate: ValidateFunction<TFieldValues>,
+): ValidateForm<TFieldValues> {
+  // Modelled after
+  // https://github.com/fabian-hiller/modular-forms/blob/main/packages/solid/src/adapters/zodForm.ts
+  // but ajv does not have async version of validate
+  return (values: PartialValues<TFieldValues>) => {
+    const formErrors: Record<string, string> = {};
+    if (!validate(values)) {
+      const ajvErrors = validate.errors ?? [];
+      for (const error of ajvErrors) {
+        const path = error.instancePath.slice(1).replace("/", ".");
+        formErrors[path] = error.message ?? "Invalid value";
+      }
+    }
+    // If child has errors, parent should have error
+    for (const key in formErrors) {
+      const path = key.split(".");
+      while (path.length > 0) {
+        path.pop();
+        const parentKey = path.join(".");
+        if (parentKey) {
+          formErrors[parentKey] = "Child properties has one or more errors";
+        }
+      }
+    }
+    return formErrors as FormErrors<TFieldValues>;
```

Comment:
This is great, but the formatting is a bit weird. Notice how the title of the next field is suddenly showing up when there's an error.

![image](https://github.com/user-attachments/assets/48bc138c-3be1-4632-824c-214fdfccbe54)

Reply:
![localhost_3000_ (4)](https://github.com/user-attachments/assets/2cfb565c-b38e-4d3f-b5c1-13ed54b4e042)

I moved the error closer to the input field. Does this help?